pykafka/producer.py

```diff
@@ -395,9 +405,26 @@ def queue_reader():
                     # surface all exceptions to the main thread
                     self.producer._worker_exception = sys.exc_info()
                     break
+            try:
```

Comment:
Yeah I agree that if it can be done without blocking `produce()` (at least for `async=True`), that would be neat. But how do you guarantee that the stopping threads aren't going to push their messages onto `_waiting_messages` too late? If that happens, I think they might sit there for ages until some other event triggers `_update`.